plugins/geojson/src/main/java/com/mapbox/mapboxsdk/plugins/geojson/GeoJsonPlugin.java

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /**
+     * AsyncTask that parse GeoJson String to {@link JSONObject}
+     */
+    private class ParseGeoJsonFromString extends AsyncTask<Void, Void, DataModel> {
+        @Override
+        protected DataModel doInBackground(Void... voids) {
+            return parseGeoJsonString(geoJson);
+        }
+
+        @Override
+        protected void onPostExecute(DataModel dataModel) {
+            super.onPostExecute(dataModel);
+            DrawOnMap(dataModel);
+            if (loadURL != null) {
+                loadURL.onLoaded();
+            }
+        }
+    }
+
+    /**
+     * AsyncTask that loading GeoJson file from special path
+     */
+    private class DrawGeoJsonFromPath extends AsyncTask<Void, Void, DataModel> {
+        @Override
+        protected void onPreExecute() {
+            if (loadFilePath != null)
+                loadFilePath.onPreLoading();
+        }
+
+        @Override
+        protected DataModel doInBackground(Void... voids) {
+            DataModel dataModel = null;
+            try {
+                InputStream inputStream = new FileInputStream(filePath);
+                dataModel = parseGeoJsonInputStream(inputStream, loadFilePath);
+            } catch (FileNotFoundException e) {
+                loadFilePath.onLoadFailed(e);
+            }
+            return dataModel;
+        }
+
+        @Override
+        protected void onPostExecute(DataModel dataModel) {
+            super.onPostExecute(dataModel);
+            if (dataModel != null) {
+                DrawOnMap(dataModel);
+                if (loadFilePath != null)
+                    loadFilePath.onLoaded();
+            }
+        }
+    }
+
+    /**
+     * AsyncTask that loading GeoJson file from Assets folder
+     */
+    private class DrawGeoJsonFromAssets extends AsyncTask<Void, Void, DataModel> {
+        @Override
+        protected void onPreExecute() {
+            if (loadFileAssets != null)
+                loadFileAssets.onPreLoading();
+        }
+
+        @Override
+        protected DataModel doInBackground(Void... voids) {
+            DataModel dataModel = null;
+            try {
+                InputStream inputStream = context.getAssets().open(fileName);
+                dataModel = parseGeoJsonInputStream(inputStream, loadFileAssets);
+            } catch (IOException e) {
+                loadFileAssets.onLoadFailed(e);
+            }
+
+            return dataModel;
+        }
+
+        @Override
+        protected void onPostExecute(DataModel dataModel) {
+            super.onPostExecute(dataModel);
+            if (dataModel != null) {
+                DrawOnMap(dataModel);
+                if (loadFileAssets != null)
+                    loadFileAssets.onLoaded();
+            }
+        }
+    }
+
+    /**
+     * @param geoJson String of GeoJson file
+     * @return DataModel list of polylines, polygons and point with bounded
+     */
+    private DataModel parseGeoJsonString(String geoJson) {
```

Comment:
This 50+ line method has multiple responsibilities and can be re factored into smaller classes that can be unit tested.